sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -137,19 +172,46 @@ private[spark] object CosmosClientCache extends BasicLoggingTrait {
     try {
       logInfo(s"-->onCleanup (${cache.size} clients)")
       val snapshot = cache.readOnlySnapshot()
+      val staleClientsInUse = ArrayBuffer[OwnerInfo]()
       snapshot.foreach(pair => {
         val clientConfig = pair._1
         val clientMetadata = pair._2
 
         if (clientMetadata.lastRetrieved.get() < Instant.now.toEpochMilli - unusedClientTtlInMs) {
-          // Only remove the client from the cache - don't purge or close!!! (because the client retrieved 15 minutes
-          // ago might still be used
-          logInfo(s"Removing client due to inactivity from the cache - ${clientConfig.endpoint}, " +
-            s"${clientConfig.applicationName}, ${clientConfig.preferredRegionsList}, ${clientConfig.useGatewayMode}, " +
-            s"${clientConfig.useEventualConsistency}")
-          cache.remove(clientConfig)
+          if (clientMetadata.refCount.get() == 0) {
+            if (clientMetadata.lastModified.get() < Instant.now.toEpochMilli - (cleanupIntervalInSeconds * 1000)) {
+              logInfo(s"Removing client due to inactivity from the cache - ${clientConfig.endpoint}, " +
+                s"${clientConfig.applicationName}, ${clientConfig.preferredRegionsList}, ${clientConfig.useGatewayMode}, " +
+                s"${clientConfig.useEventualConsistency}")
+              purgeImpl(clientConfig)
+            } else {
+              logInfo("Client has not been retrieved from the cache recently and no spark task has been using " +
+                s"it for < $cleanupIntervalInSeconds seconds. Waiting one more clean-up cycle before closing it, in " +
+                s"case newly scheduled spark tasks need it - Created: ${clientMetadata.created}, " +
+                s"LastModified ${clientMetadata.lastModified}, RefCount: ${clientMetadata.refCount}, " +
+                s"Owning Spark tasks: [${clientMetadata.owners.keys.mkString(", ")}]")
+            }
+          } else {
+            logInfo(s"Client has not been retrieved from the cache recently - Created: ${clientMetadata.created}, " +
```

Comment:
nit: Should all this be verbose and not info? Or Debug? In which scenario would the user want this as information?